core/src/main/java/org/apache/calcite/runtime/XmlFunctions.java

```diff
@@ -215,6 +238,27 @@ private static String convertNodeToString(Node node) throws TransformerException
     return writer.toString();
   }
 
+  private static Node getDocumentNode(final String xml) {
+    final DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
```

Comment:
Do we need to create a factory on every call to this method? Since the configuration does not change it may be safe to reuse the instance (per application or per thread).

Reply:
Seems reasonable: created ThreadLocal DocumentBuilderFactory